src/main/java/com/techsophy/tsf/runtime/form/service/impl/FormValidationServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+package com.techsophy.tsf.runtime.form.service.impl;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.techsophy.tsf.runtime.form.config.GlobalMessageSource;
+import com.techsophy.tsf.runtime.form.dto.*;
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.techsophy.tsf.runtime.form.constants.FormDataConstants.CHILDREN;
+import static com.techsophy.tsf.runtime.form.constants.FormModelerConstants.COMPONENTS;
+
+@Service
+@AllArgsConstructor(onConstructor_ = {@Autowired})
+public class FormValidationServiceImpl
+{
+    private ObjectMapper objectMapper;
+    private MongoTemplate mongoTemplate;
+    private GlobalMessageSource globalMessageSource;
+
+    public List<ValidationResult> validateData(FormResponseSchema formResponseSchema, FormDataSchema formData,String formId)
+    {
+        List<ValidationResult> validationResultList=new ArrayList<>();
+        Map<String, Object> components = formResponseSchema.getComponents();
+        Map<String, Object> formDataMap = formData.getFormData();
+        List<LinkedHashMap> componentsList= (List<LinkedHashMap>)components.get(COMPONENTS);
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        componentsList.forEach(x->{
+            if(x.get("type").equals("textarea"))
+            {
+                x.remove("rows");
+            }
+            if(x.get("type").equals("tree"))
+            {
+                List<LinkedHashMap> componentList= (List<LinkedHashMap>) x.get("components");
+                componentList.forEach(y->y.remove("rows"));
+                x.put("components",componentList);
+            }
+            Component  component=objectMapper.convertValue(x,Component.class);
+            validationResultList.addAll(validateComponent("",component,formDataMap,formId));
+        });
+        return validationResultList;
+    }
+
+    public void internalLoop(String prefix,Component component,List<ValidationResult> validationResultList,List<Component> componentList,Map<String,Object> data,String formId)
+    {
+        for(Component c:componentList)
+        {
+            validationResultList.addAll(validateComponent(prefix,c,component.getData(data,c.getType(),c.getLabel()),formId));
+        }
+    }
+
+    public List<ValidationResult> validateComponent(String prefix, Component component,Map<String,Object> data,String formId)
+    {
+        List<ValidationResult> validationResultList=new ArrayList<>();
+        if(component.isContainer(component)&&!component.getType().equals("address"))
+        {
+            if(!component.getType().equals("fieldset")&&!component.getType().equals("well")
```

Comment:
if you need to refer to same value again and again, there is something called variable
string componentType = component.getType()